M3 Segmented Button widget

darrenaustin opened the pull request:
Implements the [Material 3 Segmented button](https://m3.material.io/components/segmented-buttons/overview) widget.

<img width="387" alt="Screen Shot 2022-10-19 at 1 48 53 PM" src="https://user-images.githubusercontent.com/19588/196801109-75c48e65-2acf-4105-954a-a0e9793556cd.png">

Added the following new APIs:
- `SegmentedButton` - the new widget
- `ButtonSegment` - data class to describe each segment
- `SegmentedButtonTheme/Data` - it's theme

It supports single and multiple selection, as well as an option for an empty selection.

A simple example:

```dart
enum Calendar { day, week, month, year }

// In a stateful widget

  Widget build(BuildContext context) {
    return SegmentedButton<Calendar>(
      segments: const [
        ButtonSegment(value: Calendar.day, label: Text('Day')),
        ButtonSegment(value: Calendar.week, label: Text('Week')),
        ButtonSegment(value: Calendar.month, label: Text('Month')),
        ButtonSegment(value: Calendar.year, label: Text('Year')),
      ],
      selected: <Calendar>{_selected},
      onSelectionChanged: (Set<Calendar> selected) {
        setState(() {
          _selected = selected.first;
        });
      },
    );
  }
```

Fixes: #103529.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

darrenaustin committed:
Initial pass at implementing a M3 Segmented Button widget.

darrenaustin committed:
Updated to use standard outlined borders for the shape.

HansMuller commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,608 @@
+const double _kMinSegmentedButtonHeight = 28.0;
+
+/// Data describing a segment of a [SegmentedButton].
+class SegmentData<T> {
```
This seems like a somewhat generic name. SegmentedButtonSegment would be horrible, albeit more specific. Maybe just `ButtonSegment`?

HansMuller commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,608 @@
+  bool get _enabled => onSelectionChanged != null;
+
+  void _handleOnPressed(T segmentValue) {
+    if (_enabled) {
```
Might be easier to read if this was:
```dart
if (!_enabled) {
  return;
}

darrenaustin committed:
Fixed some analysis problems.

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,726 @@
+class ButtonSegment<T> {
+  /// Construct a SegmentData
+  ///
+  /// One of [icon] or [label] must be non-null.
```
This seems a little odd, since they're both Widgets.  Is it because of different default styles, or needing to apply different styles?

darrenaustin replied:
Are you commenting on the fact that they are separate widgets, or that one of them must be non-null? 

The reason they are separate is that when an item with both an icon and a label is selected, the icon is replaced with a checkbox (see the second row of the image above). In order to do this, they need to be distinct widgets.

gspencergoog replied:
I think I misread the comment.  I was thinking it was saying that only one of them could be set, and since they were both widgets, I wondered why it wasn't just one value. I see now that both can be set.

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,726 @@
+  /// [ButtonSegment.icon] if it is specified.
+  ///
+  /// Defaults to [Icons.check].
+  final IconData? selectedIcon;
```
Could this be a Widget instead?

darrenaustin replied:
Possibly. This value can also be specified in the `SegmentedButtonThemeData` and I was a little concerned about putting a Widget into a theme data structure.

darrenaustin replied:
@HansMuller do you have any issue with use a Widget in a theme data field?  I don't think we have done that before, is there a good reason not to?

HansMuller replied:
I don't think there's any reason not to. You'll have to configure the icon widget indirectly with IconTheme and you will not have control over the icon widget's size. Using a widget might be convenient for apps that just want to add a little padding. OTOH, using a widget enables a bunch of use cases, like icons that handle input, which we probably don't care about. I don't feel strongly about this one way or the other.

darrenaustin replied:
Sounds good. I switched it to a `Widget`.

darrenaustin committed:
Fixed doc issues.

HansMuller commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,726 @@
+    final OutlinedBorder disabledBorder = (resolvedDisabledBorder ?? const RoundedRectangleBorder()).copyWith(side: disabledSide);
+    final OutlinedBorder? borderShape = resolvedEnabledBorder?.copyWith(side: BorderSide.none);
+
+    final List<Widget> buttons = segments.map(buttonFor).toList();
```
NICE

HansMuller commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,726 @@
+  List<ButtonSegment<T>> get segments => _segments;
+  List<ButtonSegment<T>> _segments;
+  set segments(List<ButtonSegment<T>> value) {
+    if (_segments == value) {
```
I wonder if we should use [listEquals](https://api.flutter.dev/flutter/foundation/listEquals.html) here, since build methods are likely to rebuild the list of segments each time the selection changes.

HansMuller reviewed (commented):
This all looks good. Obviously a complete set of tests are needed.

darrenaustin committed:
Responded to review comments.

darrenaustin commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,726 @@
+    final SegmentedButtonThemeData defaults = _SegmentedButtonDefaultsM3(context);
+    final TextDirection direction = Directionality.of(context);
+    final ButtonStyle style = this.style ?? theme.style ?? defaults.style!;
+    final ButtonStyle segmentStyle = ButtonStyle(
```
Yeah, I wanted to control exactly which properties were being used on the segment buttons. Specifically the sizing and shape and side shouldn't go down to them.

darrenaustin committed:
More updates from feedback and fixed a test issue.

darrenaustin committed:
Fixed several rendering issues and added docs and tests.

HansMuller commented on packages/flutter/lib/src/material/segmented_button_theme.dart:
```diff
@@ -0,0 +1,131 @@
+/// Defines the color and border properties of [SegmentedButton] widgets.
+///
+/// Used by [SegmentedButtonTheme] to control the color and border properties
+/// of toggle buttons in a widget subtree.
```
Grep for "Toggle" and especially "ToggleButtons" in all of the files in this PR.

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+///
+/// The options are represented by segments described with [ButtonSegment]
+/// entries in the [segments] field. Each segment has a [ButtonSegment.value]
+/// that is used to indicated which segment's are selected.
```
```suggestion
/// that is used to indicate which segments are selected.
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+/// entries in the [segments] field. Each segment has a [ButtonSegment.value]
+/// that is used to indicated which segment's are selected.
+///
+/// The [selected] field is a set of the selected segments' values. This should
```
```suggestion
/// The [selected] field is a set of selected segment values. This should
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+/// The [selected] field is a set of the selected segments' values. This should
+/// be updated by the app in response to [onSelectionChanged] updates.
+///
+/// By default only a single segment can be selected (for mutually exclusive
```
```suggestion
/// By default, only a single segment can be selected (for mutually exclusive
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+/// By default only a single segment can be selected (for mutually exclusive
+/// choices). This can be relaxed with the [multiSelectionEnabled] field.
+///
+/// Like [ButtonStyleButton]s the [SegmentedButton]'s visuals can be
```
```suggestion
/// Like [ButtonStyleButton]s, the [SegmentedButton]'s visuals can be
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+/// buttons, some of the style parameters are applied to the entire segmented
+/// button, and others are used for each of the segments.
+///
+/// By default a checkmark icon is used to show selected items. To configure
```
```suggestion
/// By default, a checkmark icon is used to show selected items. To configure
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+/// this behavior, you can use the [showSelectedIcon] and [selectedIcon] fields.
+///
+/// Individual segments can be enabled or disabled with their
+/// [ButtonSegment.enabled] flag, but if [onSelectionChanged] field is null
```
```suggestion
/// [ButtonSegment.enabled] flag. If the [onSelectionChanged] field is null,
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+///   * Material Design spec: <https://m3.material.io/components/segmented-buttons/overview>
+///   * [ButtonStyle], which can be used to configure the appearance of the button.
+///   * [Radio], an alternative way to present the user with a mutually exclusive set of options.
+///   * [FilterChip], [ChoiceChip], which can be used when you need to show more then five options.
```
```suggestion
///   * [FilterChip], [ChoiceChip], which can be used when you need to show more than five options.
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+/// See also:
+///
+///   * Material Design spec: <https://m3.material.io/components/segmented-buttons/overview>
+///   * [ButtonStyle], which can be used to configure the appearance of the button.
```
Maybe mention the `[style]` field?

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+///   * [Radio], an alternative way to present the user with a mutually exclusive set of options.
+///   * [FilterChip], [ChoiceChip], which can be used when you need to show more then five options.
+class SegmentedButton<T> extends StatelessWidget {
+  /// Creates a segmented button.
```
```suggestion
  /// Creates a const [SegmentedButton].
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+class SegmentedButton<T> extends StatelessWidget {
+  /// Creates a segmented button.
+  ///
+  /// [segments] must contain at least one segment, but it is recommended
```
What does it mean to have a segmented button with only one segment?  That seems very weird, UI-wise.

guidezpl replied:
We should restrict to 2-5

w1th0utnam3 replied:
Sorry for intruding here, I'm just a Flutter user and got a notification because I subscribed to this issue, so feel free to ignore this comment. Personally I would argue against artificially restricting the number of segments as long as an arbitrary number of segments doesn't inherently break anything with the representation of the button. I only see the following problems with a large number of segments:
1. It can easily overflow or have similar issues.
2. It's against the material design guidelines.

In my opinion these points don't justify having this hard restriction which would require a user with a potentially valid use case to duplicate the entire code of this 700 line widget (plus maybe the theme data). In particular:

1. Overflow or similar issues can also occur with e.g. five very long segments.
2. There are many places in the library where a user can do stuff that is against the material design guidelines (e.g. using wrong paddings, choosing colors that don't fit according to the spec etc.). It is not a clear case to me that Flutter should enforce the material guidelines. Maybe a user has a different design concept for the entire app but the general functionality of this segmented button is still useful for them. In any case the documentation could just link to the guidelines for the suggested maximum of 5 items. I think there are other widgets that also refer to the guidelines to _suggest_ a conforming design.

For me as a developer it would appear very arbitrary if there was e.g. an `assert(segments.length < 5)` statement in the code.

guidezpl replied:
Good points, thank you. I just noticed the recommendation below for two to five. It sounds reasonable to avoid specifying an upper limit, but a single segmented button doesn't make sense UI wise. A button or chip or checkbox or switch would be more appropriate in those cases.

darrenaustin replied:
Yeah, I thought about restricting it to 5, but for the reasons that @w1th0utnam3 mentioned above I went with just the recommendation. 

As for allowing a single segment, while it would be a little weird, there is nothing technically wrong in terms of the implementation. If people feel strongly about this, we can add an assert that there are at least 2 segments.

gspencergoog replied:
Maybe just add a comment mentioning that a one-segment segmented button is probably better represented as a choice chip, no need to assert that there is more than one.

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+  /// [segments] must contain at least one segment, but it is recommended
+  /// to have two to five segments.
+  ///
+  /// If [onSelectionChanged] is null then the entire segemented button will
```
```suggestion
  /// If [onSelectionChanged] is null, then the entire [SegmentedButton] will
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+  /// be disabled.
+  ///
+  /// By default [selected] must only contain one entry. However, if
+  /// [multiSelectionEnabled] is true then [selected] can contain multiple entries.
```
```suggestion
  /// [multiSelectionEnabled] is true, then [selected] can contain multiple entries.
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+  /// Descriptions of the segments in the button.
+  final List<ButtonSegment<T>> segments;
+
+  /// Set of [ButtonSegment.value]s that indicate which [segments] are selected.
```
```suggestion
  /// The set of [ButtonSegment.value]s that indicate which [segments] are selected.
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+  /// Set of [ButtonSegment.value]s that indicate which [segments] are selected.
+  final Set<T> selected;
+
+  /// Function that is called when the selection changes.
```
```suggestion
  /// The function that is called when the selection changes.
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+  ///
+  /// The callback's parameter indicates which of the segments are selected.
+  ///
+  /// When the callback is null, the entire segmented button is disabled,
```
```suggestion
  /// When the callback is null, the entire [SegmentedButton] is disabled,
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+  /// segment, the other selected segments will stay selected. Selecting an
+  /// already selected segment will unselect it.
+  ///
+  /// If false, only one segment will be selected at a time. When a segment
```
```suggestion
  /// If false, only one segment may be selected at a time. When a segment
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+  /// is selected, any previously selected segment will be unselected.
+  final bool multiSelectionEnabled;
+
+  /// Determines if having no selected segments is allowed.
```
Maybe describe behavior for each case, and describe what the default is.

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+  ///
+  /// If false, only one segment will be selected at a time. When a segment
+  /// is selected, any previously selected segment will be unselected.
+  final bool multiSelectionEnabled;
```
Describe what the default is.

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+  ///
+  /// When the callback is null, the entire segmented button is disabled,
+  /// and will not respond to input.
+  final void Function(Set<T>)? onSelectionChanged;
```
Describe what the default is.

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+  final List<ButtonSegment<T>> segments;
+
+  /// Set of [ButtonSegment.value]s that indicate which [segments] are selected.
+  final Set<T> selected;
```
Mention that this parameter is required.

gspencergoog replied:
I would mention that it is "required" here because you might not be looking at the constructor signature when you're looking at the API doc page for this parameter.  It's in the same vein as mentioning what the default value is.

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+        assert(selected.length < 2 || multiSelectionEnabled);
+
+  /// Descriptions of the segments in the button.
+  final List<ButtonSegment<T>> segments;
```
Mention that this parameter is required.

darrenaustin replied:
But isn't that already covered by the constructor signature? Do we normally add required docs to the individual fields?

gspencergoog replied:
We often do, and we often mention that something is "optional" if it's nullable.

It's not required (to say it), though. :-)

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+  ///   * [ButtonStyle.splashFactory]
+  final ButtonStyle? style;
+
+  /// Determines if the [selectedIcon] (usually a checkmark) is displayed on
```
Maybe make a link to the [Icons.check] so it's clear which one it is.

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+  /// Determines if the [selectedIcon] (usually a checkmark) is displayed on
+  /// the selected segments.
+  ///
+  /// If true the [selectedIcon] will be displayed at the start of the segment.
```
```suggestion
  /// If true, the [selectedIcon] will be displayed at the start of the segment.
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+  /// of the segment.
+  ///
+  /// If false, then the [selectedIcon] is not used and will not be displayed
+  /// on selected segments.
```
Describe what the default is.

gspencergoog commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+  /// will be shown before the [ButtonSegment.label], replacing the
+  /// [ButtonSegment.icon] if it is specified.
+  ///
+  /// Defaults to `const Icon(Icons.check)`.
```
It would be nice if there was a link to `[Icons.check]` so they could see it in the IDE, or click to it in the docs to see what it looks like.

gspencergoog commented on packages/flutter/lib/src/material/segmented_button_theme.dart:
```diff
@@ -0,0 +1,140 @@
+
+// Examples can assume:
+// late BuildContext context;
+
```
Remove extra blank line

gspencergoog commented on packages/flutter/lib/src/material/segmented_button_theme.dart:
```diff
@@ -0,0 +1,140 @@
+/// [SegmentedButton] widgets.
+///
+/// Descendant widgets obtain the current [SegmentedButtonThemeData] object with
+/// `SegmentedButtonTheme.of(context)`. Instances of [SegmentedButtonTheme] can
```
```suggestion
/// [SegmentedButtonTheme.of]. Instances of [SegmentedButtonTheme] can
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button_theme.dart:
```diff
@@ -0,0 +1,140 @@
+/// Typically a [SegmentedButtonTheme] is specified as part of the overall
+/// [Theme] with [ThemeData.segmentedButtonTheme].
+///
+/// All [SegmentedButtonThemeData] properties are `null` by default. When null,
```
```suggestion
/// All [SegmentedButtonThemeData] properties are null by default. When null,
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button_theme.dart:
```diff
@@ -0,0 +1,140 @@
+/// [Theme] with [ThemeData.segmentedButtonTheme].
+///
+/// All [SegmentedButtonThemeData] properties are `null` by default. When null,
+/// the [SegmentedButton] compute its own default values, typically based on
```
```suggestion
/// the [SegmentedButton] computes its own default values, typically based on
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button_theme.dart:
```diff
@@ -0,0 +1,140 @@
+/// [ThemeData.iconTheme].
+@immutable
+class SegmentedButtonThemeData with Diagnosticable {
+  /// Creates the theme data that can be used override default properties
```
```suggestion
  /// Creates the [SegmentedButtonThemeData] that can be used override default properties
```

guidezpl replied:
```suggestion
  /// Creates the [SegmentedButtonThemeData] that can be used to override default properties
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button_theme.dart:
```diff
@@ -0,0 +1,140 @@
+    this.selectedIcon,
+  });
+
+  /// Overrides for the [SegmentedButton]'s default style.
```
```suggestion
  /// Overrides the [SegmentedButton]'s default style.
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button_theme.dart:
```diff
@@ -0,0 +1,140 @@
+  /// If [style] is null, then this theme doesn't override anything.
+  final ButtonStyle? style;
+
+  /// Override for [SegmentedButton.selectedIcon].
```
```suggestion
  /// Overrides the [SegmentedButton.selectedIcon] property.
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button_theme.dart:
```diff
@@ -0,0 +1,140 @@
+
+  /// Override for [SegmentedButton.selectedIcon].
+  ///
+  /// If non-null then [selectedIcon] will be used instead of default
```
```suggestion
  /// If non-null, then [selectedIcon] will be used instead of default
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button_theme.dart:
```diff
@@ -0,0 +1,140 @@
+  /// value for [Segmented.selectedIcon].
+  final Widget? selectedIcon;
+
+  /// Creates a copy of this object but with the given fields replaced with the
```
```suggestion
  /// Creates a copy of this object with the given fields replaced with the
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button_theme.dart:
```diff
@@ -0,0 +1,140 @@
+    );
+  }
+
+  /// Linearly interpolate between two segmented button themes.
```
```suggestion
  /// Linearly interpolates between two segmented button themes.
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button_theme.dart:
```diff
@@ -0,0 +1,140 @@
+/// Values specified here are used for [SegmentedButton] properties that are not
+/// given an explicit non-null value.
+class SegmentedButtonTheme extends InheritedTheme {
+  /// Creates a segmented button theme that controls visual parameters for
```
```suggestion
  /// Creates a [SegmentedButtonTheme] that controls visual parameters for
```

gspencergoog commented on packages/flutter/lib/src/material/segmented_button_theme.dart:
````diff
@@ -0,0 +1,140 @@
+  /// ```dart
+  /// SegmentedButtonThemeData theme = SegmentedButtonTheme.of(context);
+  /// ```
+  static SegmentedButtonThemeData of(BuildContext context) {
````
There should also be a related `maybeOf` for this static that returns a nullable `SegmentedButtonThemeData`. You can then implement this function in terms of that one.

Also take a look at the doc comments for the other `of` functions: you should include what happens if nothing is found, and a "See also" section that points to the `maybeOf` (and vice versa).

Is this also idiomatic with other button themes?  It seems odd that it would return the `Theme.of(context)` version if it didn't find one, instead of having the user do that fallback using `maybeOf`. I would have expected it to throw.

darrenaustin replied:
This was pretty much lifted from TextButtonThemeData.of. I will update this to match other `of` functions. We can fix the others in a separate PR.

darrenaustin replied:
And yes it is idiomatic of all the themes to fallback to the `Theme.of(context).fooTheme` in the `.of` method.

gspencergoog commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -1275,42 +1279,50 @@ class ThemeData with Diagnosticable {
   /// running on the web, otherwise it will fallback to [InkRipple].
   ///
+  /// If [brighness] is [Brightness.dark] then the default color scheme will
+  /// be either the M3 baseline dark color scheme or M2 baseline dark color
```
```suggestion
  /// be either the M3 baseline dark color scheme or the M2 baseline dark color
```

guidezpl commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -0,0 +1,770 @@
+  final bool enabled;
+}
+
+/// A Material button that allows the user to select from limited set of options.
+///
+/// Segmented buttons are used to help people select options, switch views, or
```
Please cross-reference at least from `ToggleButtons` to this class

guidezpl commented:
Great work! 

```
child: SegmentedButton<int>(
  segments: const <ButtonSegment<int>>[
    ButtonSegment<int>(value: 1, label: Text('1')),
    ButtonSegment<int>(value: 2, label: Text('2')),
    ButtonSegment<int>(value: 3, label: Text('3')),
  ],
  selected: <int>{selected},
  onSelectionChanged: (Set<int> selected) {
    selectedSegment = selected.first;
  },
),
```
was slightly unclear to me (e.g. no `setState`) Can you clarify usage by providing a Dartpad sample?

darrenaustin committed:
Updates from review comments.

darrenaustin committed:
More updates from review comments.

darrenaustin committed:
Added a sample and fixed some doc errors.

darrenaustin commented:
> Great work!

Thx.

> was slightly unclear to me (e.g. no `setState`) Can you clarify usage by providing a Dartpad sample?

I just added a example that shows both single and multi-select cases, and updated the example in the description to be a little clearer about the state handling.

HansMuller commented on examples/api/lib/material/segmented_button/segmented_button.0.dart:
```diff
@@ -0,0 +1,103 @@
+      selected: <Calendar>{_selected},
+      onSelectionChanged: (Set<Calendar> selected) {
+        setState(() {
+          _selected = selected.first;
```
It might be simpler to just carry the Set through the implementation, as in the MultipleChoice example. Also, some  naming micro suggestions.
```dart
// By default the SegmentedButton only allows one item to be
// selected at a time. The value of that item is selection.first.
Calendar selection = {Calendar.day};

 onSelectionChanged: (Set<Calendar> newSelection) {
   setState(() {
     selection = newSelection;
   }
 }
```

darrenaustin replied:
Agreed on the naming issue, but I think it makes sense to show a single value state for the single choice example as that is probably a more common use case and storing the value as a set of one value seems confusing.

HansMuller replied:
OK, thanks for considering it.

darrenaustin replied:
No worries. If you feel strongly about it, I can definitely switch it. I just thought it would be good to show how to deal with the two main uses in ways most apps probably need.

darrenaustin committed:
Updated example from review comments.

guidezpl reviewed (approved):
```
  _    ___ _____ __  __ 
 | |  / __|_   _|  \/  |
 | |_| (_ | | | | |\/| |
 |____\___| |_| |_|  |_|
                        
```

HansMuller reviewed (approved):
LGTM

gspencergoog reviewed (approved):
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)

Looks great!

darrenaustin committed:
More doc updates from review comments.

darrenaustin committed:
Fixed a doc typo

OlaiyaO commented:
I ran into an issue using the segmented button and I think there might be something that needs to be fixed about it. I think the icon and the text in each cell can only be horizontally aligned. my design calls for the icon and the text stacked vertically... The icon on top and the text below. Is there anything wrong with this idea?

TahaTesser commented:
> I ran into an issue using the segmented button and I think there might be something that needs to be fixed about it. I think the icon and the text in each cell can only be horizontally aligned. my design calls for the icon and the text stacked vertically... The icon on top and the text below. Is there anything wrong with this idea?

Can you please file a new issue with reproducible steps and code sample? Thank you!
